File: python/islands_and_treasure.py
class Solution:
    # O(n * m) time | O(n * m) space
    def islandsAndTreasure(self, grid):
        gates = []

        for r in range(len(grid)):
            for c in range(len(grid[r])):
                if grid[r][c] == 0:
                    gates.append([r,c])
        
        directions = [[0, 1], [0, -1], [1, 0], [-1, 0]]

        while gates:
            position = gates.pop()
            row = position[0]
            col = position[1]
            cost = grid[row][col]

            for dir in directions:
                newRow = row + dir[0]
                newCol = col + dir[1]

                if self.isValidPos(grid, newRow, newCol) and grid[newRow][newCol] != -1 and cost + 1 < grid[newRow][newCol]:
                    grid[newRow][newCol] = cost + 1
                    gates.append([newRow, newCol])
    
    def isValidPos(self, grid, row, col):
        return row >= 0 and row <= len(grid) - 1 and col >= 0 and col <= len(grid[row]) - 1


grid = [
  [2147483647,-1,0,2147483647],
  [2147483647,2147483647,2147483647,-1],
  [2147483647,-1,2147483647,-1],
  [0,-1,2147483647,2147483647]
]

solution = Solution()
solution.islandsAndTreasure(grid)

for r in range(len(grid)):
    vals = []
    for c in range(len(grid[r])):
        vals.append(grid[r][c])
    print(vals)

        

File: src/main/java/august_2025/IslandsAndTreasure.java
package august_2025;

import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class IslandsAndTreasure {

    static int INF = Integer.MAX_VALUE;

    public static void main(String[] args) {
        int[][] grid = {
                {INF, -1, 0, INF},
                {INF, INF, INF, -1},
                {INF, -1, INF, -1},
                {0, -1, INF, INF}
        };

        IslandsAndTreasure islandsAndTreasure = new IslandsAndTreasure();
        islandsAndTreasure.islandsAndTreasure(grid);

        System.out.println();

    }

    // O(n * m) time | O(n * m) space
    public void islandsAndTreasure(int[][] grid) {

        Queue<int[]> queue = new LinkedList<>();

        for (int r = 0; r < grid.length; r++) {
            for (int c = 0; c < grid[r].length; c++) {
                if (grid[r][c] == 0) {
                    queue.add(new int[] {r, c});
                }
            }
        }

        int[][] directions = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};

        while (!queue.isEmpty()) {
            int[] current = queue.poll();
            int x = current[0];
            int y = current[1];
            int cost = grid[x][y];

            for (int[] dir : directions) {
                int newX = x + dir[0];
                int newY = y + dir[1];

                if (isValidPos(grid, newX, newY) && isLand(grid, newX, newY)) {
                    int newCost = cost + 1;
                    if (newCost < grid[newX][newY]) {
                        grid[newX][newY] = newCost;
                        queue.add(new int[] {newX, newY});
                    }
                }
            }
        }

    }

    private boolean isLand(int[][] grid, int x, int y) {
        return grid[x][y] != -1;
    }

    private boolean isValidPos(int[][] grid, int x, int y) {
        return x >= 0 && x <= grid.length - 1 && y >= 0 && y <= grid[x].length - 1;
    }

}
